test/modules/test_nf4_linear.py

```diff
@@ -0,0 +1,74 @@
+import logging
+import unittest
+
+import torch
+from torch import nn
+from torch.testing._internal.common_utils import TestCase
+from torchao.modules import FrozenNF4Linear
+from torchao.dtypes.nf4tensor import NF4Tensor
+
+bnb_available = False
+
+try:
+    import bitsandbytes as bnb
+    bnb_available = True
+except ImportError:
+    pass
+
+logging.basicConfig(
+    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s", level=logging.INFO
+)
+
+class TestNF4Linear(TestCase):
+    """
+    Test torchao.modules.NF4Linear
+    """
+    def test_bias_unsupported(self):
+        with self.assertRaisesRegex(RuntimeError, "does not currently support biases"):
+            _ = FrozenNF4Linear(1, 1, bias=True)
+
+    def test_non_bf16_unsupported(self):
+        with self.assertRaisesRegex(RuntimeError, "only supported with bf16"):
+            _ = FrozenNF4Linear(1, 1)
+
+    def test_frozen_nf4_linear(self):
+        nf4_linear = FrozenNF4Linear(512, 512, device='cpu', dtype=torch.bfloat16)
+        self.assertTrue(isinstance(nf4_linear.weight, NF4Tensor))
+        self.assertEqual(torch.bfloat16, nf4_linear.weight.get_original_weight().dtype)
+
+    def test_output_bf16(self):
+        # Test to ensure W4 A16 produces A16
+        nf4_linear = FrozenNF4Linear(512, 512, device='cpu', dtype=torch.bfloat16)
+        inp = torch.randn(2, 512, dtype=torch.bfloat16, requires_grad=True)
+        out = nf4_linear(inp)
+        assert out.dtype == torch.bfloat16
+
+    def test_backward_bf16(self):
+        # Test to ensure backward pass gives activation a bf16 gradient and no gradient
+        # to the linear's weight, as it is frozen.
+        nf4_linear = FrozenNF4Linear(512, 512, device='cpu', dtype=torch.bfloat16)
+        inp = torch.randn(2, 512, dtype=torch.bfloat16, requires_grad=True)
+        nf4_linear(inp).sum().backward()
+        assert inp.grad is not None and inp.grad.dtype == torch.bfloat16
+        assert nf4_linear.weight.grad is None
+
+
+    def _build_bnb_linear(self, input_weight):
+        assert bnb_available, "Needs bitsandbytes support"
+        param = bnb.nn.Params4bit(input_weight, requires_grad=False, quant_type="nf4")
+        bnb_linear = bnb.nn.LinearNF4(input_weight.size(0), input_weight.size(1), bias=False)
+        bnb_linear.weight = param
+        bnb_linear.cuda()
+        return bnb_linear
+
+    @unittest.skipIf(not bnb_available, "Need bnb availble")
+    def test_fwd_backward_bnb_parity(self):
+        nf4_linear = FrozenNF4Linear(512, 512, device='cuda', dtype=torch.bfloat16)
+        orig_weight = nf4_linear.weight.get_original_weight().clone().detach()
+        bnb_nf4_linear = self._build_bnb_linear(input_weight=orig_weight)
+
+        inp = torch.randn(2, 512, dtype=torch.bfloat16, device='cuda')
+        self.assertEqual(nf4_linear(inp).sum(), bnb_nf4_linear(inp).sum())
```

Comment:
IMO this is kinda a weird test I think reconstruction accuracy is better;
- compared to original: https://github.com/drisspg/transformer_nuggets/blob/f05afad68ad9086d342268f46a7f344617a02314/test/test_qlora.py#L45
- compared against bnb: "make sure we arent worse"
https://github.com/drisspg/transformer_nuggets/blob/f05afad68ad9086d342268f46a7f344617a02314/test/test_qlora.py#L65

Reply:
Will def add reconstruction accuracy test. Curious why it's not as valuable to test exact parity w/BNB though?